src/Razor/test/Microsoft.AspNetCore.Razor.LanguageServer.Test/ProjectConfigurationStateSynchronizerTest.cs

```diff
@@ -526,13 +740,86 @@ public async Task ProjectConfigurationFileChanged_RemoveThenAdd_OnlyAdds()
+           "/path/other/obj/project.razor.json",
+           "path/other/project.csproj",
+           RazorConfiguration.Default,
+           rootNamespace: "TestRootNamespace",
+           displayName: "project",
+           ProjectWorkspaceState.Create(LanguageVersion.CSharp5),
+           documents: []);
+        var intermediateOutputPath2 = FilePathNormalizer.GetNormalizedDirectoryName(projectInfo2.SerializedFilePath);
+        var projectKey2 = TestProjectKey.Create(intermediateOutputPath2);
+
+        var projectServiceMock = new StrictMock<IRazorProjectService>();
+        projectServiceMock
+            .Setup(p => p.UpdateProjectAsync(
+                projectKey1,
+                It.IsAny<RazorConfiguration>(),
+                It.IsAny<string>(),
+                It.IsAny<string>(),
+                It.IsAny<ProjectWorkspaceState>(),
+                It.IsAny<ImmutableArray<DocumentSnapshotHandle>>(),
+                It.IsAny<CancellationToken>()))
+            .Returns(Task.CompletedTask);
+        projectServiceMock
+            .Setup(p => p.AddProjectAsync(
+                projectInfo2.FilePath,
+                intermediateOutputPath2,
+                It.IsAny<RazorConfiguration>(),
+                projectInfo2.RootNamespace,
+                projectInfo2.DisplayName,
+                It.IsAny<CancellationToken>()))
+            .ReturnsAsync(projectKey2);
+        projectServiceMock
+            .Setup(p => p.UpdateProjectAsync(
+                projectKey2,
+                It.IsAny<RazorConfiguration>(),
+                It.IsAny<string>(),
+                It.IsAny<string>(),
+                It.IsAny<ProjectWorkspaceState>(),
+                It.IsAny<ImmutableArray<DocumentSnapshotHandle>>(),
+                It.IsAny<CancellationToken>()))
+            .Returns(Task.CompletedTask);
+
+        using var synchronizer = GetSynchronizer(projectServiceMock.Object);
+        var synchronizerAccessor = synchronizer.GetTestAccessor();
+
+        var removedArgs = new ProjectConfigurationFileChangeEventArgs(projectInfo1.SerializedFilePath, RazorFileChangeKind.Removed, CreateDeserializer(projectInfo1));
+        var addedArgs = new ProjectConfigurationFileChangeEventArgs(projectInfo2.SerializedFilePath, RazorFileChangeKind.Added, CreateDeserializer(projectInfo2));
+
+        // Act
+        synchronizer.ProjectConfigurationFileChanged(removedArgs);
+        synchronizer.ProjectConfigurationFileChanged(addedArgs);
+
+        await synchronizerAccessor.WaitUntilCurrentBatchCompletesAsync();
+
+        // Assert
+        projectServiceMock.VerifyAll();
+    }
+
     private TestProjectConfigurationStateSynchronizer GetSynchronizer(IRazorProjectService razorProjectService)
-        => new(razorProjectService, LoggerFactory, TestLanguageServerFeatureOptions.Instance, TimeSpan.FromMilliseconds(5));
+        => new(razorProjectService, LoggerFactory, TestLanguageServerFeatureOptions.Instance, TimeSpan.FromMilliseconds(50));
```

Comment:
Why does this need to be 50 milliseconds? Is there a race?

Reply:
Not a race that leads to a bug, but on my laptop on battery the test where I add half a dozen changed events would sometimes end up processing two batches, which means `UpdateProjectAsync` would be called one more time than we expect. Though to be fair, the number of changed events I'm adding is a little pathological :)

Could just directly test the process method and avoid the queue for some tests though, since the event handler code is simpler now anyway.